### PR TITLE
[diffusion] Add Qwen-Image ModelOpt NVFP4 support

### PR DESCRIPTION
**Pull request by BBuf**
## Summary

The NVFP4 checkpoints used for this PR were produced locally on the B200 validation node from NVIDIA ModelOpt PR #1706, branch `qwen-image-svdquant-nvfp4-pr` (`31db2397`). I used ModelOpt's diffusers quantization entrypoint (`examples/diffusers/quantization/quantize.py`) with the `qwen-image` recipe, BF16 model loading, `--format fp4`, and the default `max` quantization algorithm. The calibration set was `/data/bbuf/qwen_image_nvfp4/prompts/qwen_calib_8.txt`, which produced 8 calibration batches in the export logs.

The four hosted checkpoints uploaded for this PR are full Hugging Face-style pipeline directories under `lmsys/*`:

- `lmsys/qwen-image-modelopt-nvfp4-sglang` from `Qwen/Qwen-Image`
- `lmsys/qwen-image-2512-modelopt-nvfp4-sglang` from `Qwen/Qwen-Image-2512`
- `lmsys/qwen-image-edit-modelopt-nvfp4-sglang` from `Qwen/Qwen-Image-Edit`
- `lmsys/qwen-image-edit-2511-modelopt-nvfp4-sglang` from `Qwen/Qwen-Image-Edit-2511`

The local B200 export directories were kept under `/data/bbuf/qwen_image_nvfp4/ckpts/` before upload. `Qwen-Image-Edit-2509` was validated locally but is not part of the hosted checkpoint set for this PR.

ModelOpt PR #1706 exposes Qwen Image through the `qwen-image` model type, so the edit variants were exported through the same recipe with the base model path overridden to the edit repos. The Qwen recipe quantized only `transformer_blocks`, excluded the first two and last two transformer blocks, used NVFP4 block size 16, disabled attention quantizers through the Qwen filter, fused QKV amax groups during export, and saved the resulting packed NVFP4 transformer plus the unquantized pipeline components into the local HF directories above.

This PR adds Qwen-Image ModelOpt NVFP4 loading support for SGLang Diffusion, based on the draft support in #28557 and validated with public `lmsys/*` checkpoints exported from NVIDIA ModelOpt PR #1706.

Changes:
- Register the Qwen-Image NVFP4 full-model repo layout for native diffusion loading.
- Canonicalize Qwen modulation tensor names such as `img_mod.1` / `txt_mod.1` when reconstructing ModelOpt NVFP4 exclude modules.
- Add one B200 diffusion 50-step consistency case for `lmsys/qwen-image-2512-modelopt-nvfp4-sglang`, backed by a new `ci-data` GT image.
- Document the hosted `lmsys/*` checkpoint workflow and B200 performance caveat in `docs` and `docs_new`, including Qwen-Image-2512 and Qwen-Image-Edit variant validation.

The hosted `lmsys/*` checkpoints are now public and readable from CI. This PR enables one representative B200 NVFP4 50-step consistency test for `lmsys/qwen-image-2512-modelopt-nvfp4-sglang`; the other hosted Qwen Image NVFP4 repos are documented but not added to the B200 shard to keep runtime bounded.

## B200 validation

Environment:
- 8x NVIDIA B200 node, SGLang native diffusion backend.
- Base models: `Qwen/Qwen-Image`, `Qwen/Qwen-Image-2512`, `Qwen/Qwen-Image-Edit`, `Qwen/Qwen-Image-Edit-2509`, and `Qwen/Qwen-Image-Edit-2511`.
- Quantized checkpoints: public `lmsys/*` ModelOpt PR #1706 exports using `--format fp4 --quant-algo max`; Edit variants were exported through a local wrapper because PR #1706 currently exposes only the `qwen-image` model type.
- No diffusers fallback markers in representative NVFP4 logs.
- Transformer memory reduced from roughly 38.05 GB BF16 to 13.69 GB NVFP4 at load.

### Image quality spot-check (BF16 vs NVFP4, model-config steps)

The following samples were generated on B200 with the same prompt, seed, resolution, and each model config's default sampling steps. The commands did not pass `--num-inference-steps`.

- `Qwen-Image`: BF16 `Qwen/Qwen-Image` vs `lmsys/qwen-image-modelopt-nvfp4-sglang`, `infer_steps=50`
- `Qwen-Image-2512`: BF16 `Qwen/Qwen-Image-2512` vs `lmsys/qwen-image-2512-modelopt-nvfp4-sglang`, `infer_steps=50`
- `Qwen-Image-Edit`: BF16 `Qwen/Qwen-Image-Edit` vs `lmsys/qwen-image-edit-modelopt-nvfp4-sglang`, `infer_steps=50`
- `Qwen-Image-Edit-2511`: BF16 `Qwen/Qwen-Image-Edit-2511` vs `lmsys/qwen-image-edit-2511-modelopt-nvfp4-sglang`, `infer_steps=40`

### B200 benchmark rerun with the diffusion benchmark/profile skill command (updated 2026-06-24)

The Qwen-Image 1024x1024 rows below were rerun on B200 with the command shape used by `python/sglang/multimodal_gen/.claude/skills/sglang-diffusion-benchmark-profile`: native SGLang backend, seed 42, 1024x1024, `--save-output`, `--warmup`, and `--enable-torch-compile`. The commands do not pass `--num-inference-steps`, so Qwen-Image uses its model-config default `infer_steps=50`. Warmup is excluded by the perf dump, but the measured request can still include remaining torch.compile specialization in the first denoise steps; `Avg step (all 50)` is the end-to-end denoise average for the standardized command, while `Tail avg` excludes the first two compile-heavy denoise steps. The default NVFP4 FP4 GEMM path in this run is FlashInfer TRT-LLM (`trtllm_fp4_gemm_runner`).

Commands:

```bash
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=python FLASHINFER_DISABLE_VERSION_CHECK=1 \
sglang generate \
  --backend=sglang \
  --model-path=Qwen/Qwen-Image \
  --prompt="A futuristic cyberpunk city at night, neon lights reflecting on wet streets" \
  --seed=42 \
  --width=1024 \
  --height=1024 \
  --save-output \
  --warmup \
  --enable-torch-compile \
  --perf-dump-path /home/sglang-omni/bbuf/qwen_nvfp4_pr28928_rebench_20260624_results/benchmarks/bf16_qwen_image_1024_compile.json

CUDA_VISIBLE_DEVICES=0 PYTHONPATH=python FLASHINFER_DISABLE_VERSION_CHECK=1 \
sglang generate \
  --backend=sglang \
  --model-path=lmsys/qwen-image-modelopt-nvfp4-sglang \
  --prompt="A futuristic cyberpunk city at night, neon lights reflecting on wet streets" \
  --seed=42 \
  --width=1024 \
  --height=1024 \
  --save-output \
  --warmup \
  --enable-torch-compile \
  --perf-dump-path /home/sglang-omni/bbuf/qwen_nvfp4_pr28928_rebench_20260624_results/benchmarks/nvfp4_qwen_image_1024_compile.json
```

| Model | Resolution / steps | Backend | Denoise | Avg step (all 50) | Median step | Tail avg (excl. first 2) | Peak reserved |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Qwen-Image | 1024x1024 / 50 | BF16 | 50.91 s | 1018.2 ms | 139.8 ms | 139.0 ms | 46.8 GB |
| Qwen-Image | 1024x1024 / 50 | NVFP4 `flashinfer_trtllm` | 62.91 s | 1258.1 ms | 1111.2 ms | 1157.9 ms | 22.3 GB |

Takeaway from the standardized 50-step, torch-compile-enabled 1024x1024 run: NVFP4 is about 23.6% slower than BF16 for Qwen-Image denoising on this B200 setup, while reducing peak reserved memory by about 52.4%. This supersedes the earlier short-step 1024x1024 timing rows, which used explicit 8-step smoke commands and are not comparable to the benchmark/profile skill command. The earlier 2048x2048 and edit-model measurements were bring-up spot checks under shorter commands and should be treated as directional validation only, not as the standardized benchmark table.

Profiler traces for the same 1024x1024 setup were generated by adding `--profile` and `SGLANG_DIFFUSION_TORCH_PROFILER_DIR=...`; both traces passed gzip integrity checks on the B200 node. The trace files are large because SGLang diffusion profiling records CPU+CUDA events with shapes and stacks: BF16 is 2.2 GB compressed and NVFP4 is 396 MB compressed.

## Profiler

- main:

e962f4a4-3a4d-4b53-b3e5-b1e44acb0421

The red box means `nn.Linear`

- pr:

cb608baf-34a5-43ce-a369-b6c56798c093

The green box means quant overhead.

## Test Plan

On the B200 container:

```bash
PYTHONPATH=python python -m py_compile \
  python/sglang/multimodal_gen/registry.py \
  python/sglang/multimodal_gen/runtime/utils/quantization_utils.py \
  python/sglang/multimodal_gen/test/server/gpu_cases.py \
  python/sglang/multimodal_gen/test/server/testcase_configs.py
```

Model-backed validation on B200:
- Exported `Qwen/Qwen-Image`, `Qwen/Qwen-Image-2512`, `Qwen/Qwen-Image-Edit`, `Qwen/Qwen-Image-Edit-2509`, and `Qwen/Qwen-Image-Edit-2511` NVFP4 checkpoints with ModelOpt PR #1706.
- Uploaded the four hosted checkpoints listed above to `lmsys/*` and verified each public repo has `README.md`, `model_index.json`, and safetensors files through `HfApi.model_info`.
- Loaded the hosted `lmsys/qwen-image-2512-modelopt-nvfp4-sglang` checkpoint directly through `--model-path` using the native SGLang diffusion backend.
- Confirmed representative NVFP4 logs contain `Built NVFP4 quant config` and native `Loaded transformer` markers without diffusers fallback markers.
- Ran the consistency and perf runs listed above.
- Re-ran the Qwen-Image 1024x1024 benchmark with the diffusion benchmark/profile skill command shape: 50 model-config steps, `--warmup`, and `--enable-torch-compile`; BF16 denoise was 50.91s and NVFP4 denoise was 62.91s.
- Generated BF16 and NVFP4 torch profiler traces for the same 1024x1024 setup with `--profile`; both `.trace.json.gz` files passed gzip integrity checks on the B200 node.
- Added the 1-GPU Qwen-Image-2512 NVFP4 GT image in `sgl-project/ci-data` PR #23 and pinned `SGL_TEST_FILES_CI_DATA_REVISION` to `187b8fdc5f49c63d965a1b80528a27ecca9067b2`.

B200 CI consistency case added in this PR:

```bash
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=python pytest -q \
  python/sglang/multimodal_gen/test/server/test_server_b200.py \
  -k qwen_image_2512_modelopt_nvfp4_t2i -s
```

Result: `1 passed, 4 deselected` in 114.91s. The hosted Qwen-Image-2512 NVFP4 case loaded from HF cache, generated a 768x768 image with 50 denoise steps, fetched the GT image through `sgl-project/ci-data@187b8fdc5f49c63d965a1b80528a27ecca9067b2` (ci-data PR #23), and passed consistency with `clip=1.0000`, `ssim=1.0000`, `psnr=inf`, and `mean_abs_diff=0.0000`. The same run reported `avg_denoise=326.79ms`, `median_denoise=326.39ms`.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #28167508009](https://github.com/sgl-project/sglang/actions/runs/28167508009)
Latest PR Test (Extra): :white_check_mark: [Run #28167507920](https://github.com/sgl-project/sglang/actions/runs/28167507920)

### CODE REVIEW

**Comment by BBuf** [comment]
Addressed the review suggestion for the Qwen modulation exclude canonicalization helper in `0dd50bd58b` by switching to tuple `endswith(...)` + `removesuffix(...)`.

Revalidated on the B200 container at the same head:
- `PYTHONPATH=python pytest -q python/sglang/multimodal_gen/test/unit/test_transformer_quant.py::TestTransformerQuantHelpers::test_nvfp4_inference_canonicalizes_qwen_modulation_excludes -q`
- `PYTHONPATH=python pytest -q python/sglang/multimodal_gen/test/unit/test_transformer_quant.py -q` -> 16 passed

**Comment by BBuf** [comment]
CI note after the latest rerun (`f851d4a20f`): the only completed failure so far is `PR Test (AMD) / multimodal-gen-test-2-gpu-amd (linux-mi325-2gpu-sglang, 1)`, and it looks unrelated to this Qwen Image NVFP4 change.

The failing cases are existing AMD diffusion scenarios, not the new Qwen NVFP4 path:
- `flux2_modelopt_fp8_tp2_t2i`: ROCm/modelopt FP8 path fails in `torch._scaled_mm` with `HIPBLAS_STATUS_NOT_SUPPORTED`.
- `ideogram4_fp8_tp2_t2i`: HF download fails with gated repo 403 for `ideogram-ai/ideogram-4-fp8`.

The PR diff only touches Qwen NVFP4 registration, the Qwen modulation exclude canonicalization helper, its unit tests, and docs. The B200 multimodal-gen job is still queued with no runner assigned (`4-gpu-b200`).

File: docs_new/docs/sglang-diffusion/quantization.mdx [hunk 1]
```diff
@@ -408,8 +466,15 @@ sglang generate \
   intentionally want a non-default `transformer_2`.
 - On Blackwell, the diffusion ModelOpt NVFP4 path defaults to FlashInfer
   TensorRT-LLM FP4 GEMM (`flashinfer_trtllm`).
-- Direct `--model-path` loading is a compatibility path for FLUX.2 NVFP4-style
-  repos or local directories.
+- The published Qwen Image NVFP4 exports keep the `img_mod`/`txt_mod`
```
Comment (mickqian) [inline docs_new/docs/sglang-diffusion/quantization.mdx:469]: should we update the doc/cookbook for qwen-image too?

Comment (BBuf) [reply docs_new/docs/sglang-diffusion/quantization.mdx:469]: done. https://github.com/sgl-project/sglang/pull/28928/changes/4e3f486f08ac8663d2f780f4dd07f45c417141cd#diff-158956da3bb6720a80e6ab9cd368ef3a69309dbbb2d996717e2c6b0bccf96e26

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/28141414355/job/83339498960?pr=28928

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/28167508009/job/83423179724?pr=28928

**Comment by BBuf** [comment]
Merged with all ci green.
